### PR TITLE
[DateRangePicker] Fix `currentMonthCalendarPosition` not scrolling to future sibling

### LINKED ISSUE #14418: [pickers][DateRangeCalendar] Using a controlled `DateRangeCalendar` with `currentMonthCalendarPosition` breaks scroll behaviour
### Steps to reproduce

Link to live example: https://codesandbox.io/p/sandbox/muddy-cherry-x8jh2g

Steps:
1. Render at `DateRangeCalendar` with the `currentMonthCalendarPosition` with a value other than `1`
2. Feed a `value` to the component so that there's a "current" calendar. (ex a couple of days in August 2024)
3. Change the `value` so that the next month becomes "current" (ex a couple of days in September 2024)

### Current behavior

The "current" calendar never changes, lingering on the previously active calendar (August in my example). Manually scrolling to the next month reveals the value is correctly interpreted, however not scrolled to. This is only true for sibling months (when `calendars={2}`)

### Expected behavior

The component should scroll to the "next" month so that my selection is always visible.

### Context

In our application we have a preference to show the current month on the right, so we used `currentMonthCalendarPosition` to achieve this. We also have a number of shortcuts such as (last month, last week, this month etc) which users use to quickly make selections.

I strongly suspect the reason the current behaviour is how it is, because `currentMonthCalendarPosition` isn't considered when scrolling. If `currentMonthCalendarPosition` was 1, the current behavior would make complete sense, since you'd only scroll if the "previous" month suddently became selected, not the "next" one.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  System:
    OS: macOS 14.6.1
  Binaries:
    Node: 20.15.0 - ~/.nvm/versions/node/v20.15.0/bin/node
    npm: 10.7.0 - ~/.nvm/versions/node/v20.15.0/bin/npm
    pnpm: 7.5.2 - /usr/local/bin/pnpm
  Browsers:
    Chrome: 128.0.6613.113 < Using this for development
    Edge: Not Found
    Safari: 17.6
  npmPackages:
    @emotion/react: ^11.11.4 => 11.11.4 
    @emotion/styled: ^11.11.5 => 11.11.5 
    @mui/base:  5.0.0-beta.40 
    @mui/core-downloads-tracker:  5.16.4 
    @mui/material: ^5.15.20 => 5.16.4 
    @mui/private-theming:  5.16.6 
    @mui/styled-engine:  5.16.6 
    @mui/system:  5.16.7 
    @mui/types:  7.2.15 
    @mui/utils:  5.16.6 
    @mui/x-charts:  7.14.0 
    @mui/x-charts-vendor:  7.14.0 
    @mui/x-data-grid:  7.14.0 
    @mui/x-data-grid-premium: ^7.13.0 => 7.14.0 
    @mui/x-data-grid-pro: ^7.7.1 => 7.14.0 
    @mui/x-date-pickers:  7.7.1 
    @mui/x-date-pickers-pro: ^7.7.1 => 7.7.1 
    @mui/x-internals:  7.14.0 
    @mui/x-license:  7.7.1 
    @types/react: ^18.2.48 => 18.3.3 
    react: ^18.2.0 => 18.3.1 
    react-dom: ^18.2.0 => 18.3.1 
    typescript: ^5.5.2 => 5.5.2 
```
</details>

**Search keywords**: currentMonthCalendarPosition, calendar picker

### PR DESCRIPTION
**Pull request by GMchris**
fixes #14418

### CODE REVIEW

**Review by LukasTy — APPROVED** [review]
Thank you for noticing and handling this issue! 🙏 

I took the liberty to add a test asserting this behavior.

Your changes seem to impact the commercially licensed code. For any changes of this nature, we require contributors to sign the MUI’s Contributor License Agreement (CLA).

However, this change is looks like a pretty trivial bugfix that we could accept without the CLA process if we both agree that the fix is trivial: https://www.notion.so/mui-org/CLA-Contributor-License-Agreement-92ece655b1584b10b00e4de9e67eedb0?pvs=4#ede845886e494cb4b802a67563152130.
WDYT?
If you wish to go through the CLA process, please let me know and follow [this guide](https://www.notion.so/mui-org/CLA-Contributor-License-Agreement-92ece655b1584b10b00e4de9e67eedb0?pvs=4#a6a4dee106a8423ea2fededcb26025d6). 😉

**Comment by GMchris** [comment]
Hey, @LukasTy , I've followed the CLA guide and submitted both form and email.

File: packages/x-date-pickers/src/DateCalendar/tests/localization.DateCalendar.test.tsx [hunk 1]
```diff
@@ -39,8 +39,10 @@ describe('<DateCalendar /> - localization', () => {
           'Sunday',
         );
 
-        setProps({
-          adapterLocale: adapterName === 'date-fns' ? fr : 'fr',
+        act(() => {
```
[the remark this answers is not in the document]
Comment (LukasTy) [reply packages/x-date-pickers/src/DateCalendar/tests/localization.DateCalendar.test.tsx]: Thanks for bringing this up. 👍 🙏 
I have reverted that particular commit. 😉

**Comment by oliviertassinari** [comment]
@LukasTy Two side thoughts:

1. When requesting for a CLA, to add the `CLA: required` label to prevent someone else from merging without noticing. Now, to be fair, the fix is really simple, there isn't much IP to get, so we could have almost merged without.
2. https://mui.com/x/react-date-pickers/date-range-calendar/#choose-the-months-to-render I have my doubts about this `currentMonthCalendarPosition`  props, I would almost better see a prop that is a date that represents the current month displayed. I mean, we have https://github.com/mui/mui-x/blob/dd98382e7b1657235c12c1334e44a68fa003bd4a/packages/x-date-pickers-pro/src/DateRangeCalendar/DateRangeCalendar.types.ts#L98 but shouldn't people be able to use -10, +10 too?

**Comment by LukasTy** [comment]
Thank you for the insights @oliviertassinari.

> 1. When requesting for a CLA, to add the `CLA: required` label to prevent someone else from merging without noticing. Now, to be fair, the fix is really simple, there isn't much IP to get, so we could have almost merged without.

Thanks, next time I will either not approve the PR or add the necessary label until the CLA topic is resolved.
This is the second time we have had a "trivial"-like change and previously agreed to merge without signing. 😉 


> 2\. https://mui.com/x/react-date-pickers/date-range-calendar/#choose-the-months-to-render I have my doubts about this `currentMonthCalendarPosition`  props, I would almost better see a prop that is a date that represents the current month displayed. I mean, we have https://github.com/mui/mui-x/blob/dd98382e7b1657235c12c1334e44a68fa003bd4a/packages/x-date-pickers-pro/src/DateRangeCalendar/DateRangeCalendar.types.ts#L98
>      but shouldn't people be able to use -10, +10 too?

It can be -10 or +10 if someone needs it, just like anyone is free to specify whatever amount of `calendars` they want (except for overriding the TS error or PropTypes warning). 😉 
We have previously had a `currentMonth` prop, but that was replaced with `referenceDate`.
What use case do you see for `-10` or `+10`? 🤔

**Comment by oliviertassinari** [comment]
@LukasTy Ah, ok, so `referenceDate` and `currentMonthCalendarPosition` are meant to be used at the same time. This makes sense. 

The thing in my mind was about: how can a developer control the view displayed in the pickers? react-day-pickers allows people to https://daypicker.dev/docs/navigation#onmonthchange. A month prop sounds too specific though, I would expect a date prop that the component ensures is "visible" in the UI.

`referenceDate` is close but seems to be only the default value.

**Comment by LukasTy** [comment]
> The thing in my mind was about: how can a developer control the view displayed in the pickers? react-day-pickers allows people to https://daypicker.dev/docs/navigation#onmonthchange. A month prop sounds too specific though, I would expect a date prop that the component ensures is "visible" in the UI.
> 
> `referenceDate` is close but seems to be only the default value.

Great point, thank you for bringing up this possible need. 👍 
There is already an issue for it (https://github.com/mui/mui-x/issues/10869), I have added specific priority to it so that it does not get lost. 👍 
We might have overlooked this need with the `referenceDate` implementation. 🤔

**Pull request merged**